pkg/proxy/index/index_controller.go

```diff
@@ -0,0 +1,171 @@
+		AddFunc: func(obj interface{}) {
+			ws := obj.(*tenancyv1alpha1.ClusterWorkspace)
+			c.lock.RLock()
+			got := c.workspaceShardNames[logicalcluster.From(ws).Join(ws.Name)]
+			c.lock.RUnlock()
+
+			if expected := ws.Status.Location.Current; got != expected {
+				c.lock.Lock()
+				defer c.lock.Unlock()
+				c.workspaceShardNames[logicalcluster.From(ws).Join(ws.Name)] = expected
+			}
+		},
+		UpdateFunc: func(old, obj interface{}) {
+			ws := obj.(*tenancyv1alpha1.ClusterWorkspace)
+
+			c.lock.RLock()
+			got := c.workspaceShardNames[logicalcluster.From(ws).Join(ws.Name)]
+			c.lock.RUnlock()
+
+			if expected := ws.Status.Location.Current; got != expected {
+				c.lock.Lock()
+				defer c.lock.Unlock()
+				c.workspaceShardNames[logicalcluster.From(ws).Join(ws.Name)] = expected
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			if final, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+				obj = final.Obj
+			}
+			ws := obj.(*tenancyv1alpha1.ClusterWorkspace)
+
+			c.lock.Lock()
+			defer c.lock.Unlock()
+			delete(c.workspaceShardNames, logicalcluster.From(ws).Join(ws.Name))
+		},
+	})
+
+	workspaceShardInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			shard := obj.(*tenancyv1alpha1.ClusterWorkspaceShard)
+			c.lock.RLock()
+			got := c.shardBaseURLs[shard.Name]
+			c.lock.RUnlock()
+
+			if expected := shard.Spec.BaseURL; got != expected {
+				c.lock.Lock()
+				defer c.lock.Unlock()
+				c.shardBaseURLs[shard.Name] = expected
+			}
+		},
+		UpdateFunc: func(old, obj interface{}) {
+			shard := obj.(*tenancyv1alpha1.ClusterWorkspaceShard)
+			c.lock.RLock()
+			got := c.shardBaseURLs[shard.Name]
+			c.lock.RUnlock()
+
+			if expected := shard.Spec.BaseURL; got != expected {
+				c.lock.Lock()
+				defer c.lock.Unlock()
+				c.shardBaseURLs[shard.Name] = expected
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			if final, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+				obj = final.Obj
+			}
+			shard := obj.(*tenancyv1alpha1.ClusterWorkspaceShard)
+
+			c.lock.Lock()
+			defer c.lock.Unlock()
+			delete(c.shardBaseURLs, shard.Name)
+		},
+	})
+
+	return c, nil
+}
+
+// Controller watches ClusterWorkspaceShards on the root shard, and then starts informers
+// for every ClusterWorkspaceShards, watching the ClusterWorkspaces on them. It then
+// updates the workspace index, which maps logical clusters to shard URLs.
+type Controller struct {
+	workspaceIndexer cache.Indexer
+	workspaceLister  tenancylister.ClusterWorkspaceLister
+
+	workspaceShardIndexer cache.Indexer
+	workspaceShardLister  tenancylister.ClusterWorkspaceShardLister
+
+	shardInformersLock sync.RWMutex
+	shardInformers     map[string]tenancyinformer.ClusterWorkspaceInformer
+
+	lock                sync.RWMutex
+	workspaceShardNames map[logicalcluster.Name]string
+	shardBaseURLs       map[string]string
+}
+
+func (c *Controller) Start(ctx context.Context) {
```

Comment:
what's the purpose of this function? (seems to be empty)

Reply:
following the controller pattern.